File: dracpond/src/main/java/cc/chordflower/dracpond/application/config/base/BaseDatabase.java
package cc.chordflower.dracpond.application.config.base;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import jakarta.validation.constraints.DecimalMax;
import jakarta.validation.constraints.DecimalMin;

import java.io.Serializable;
import java.math.BigInteger;
import java.util.Optional;


/**
 * Definitions related to the arango database
 */
@JsonInclude( JsonInclude.Include.NON_NULL )
@JsonPropertyOrder( {
  "host",
  "port",
  "secure",
  "username",
  "password"
} )
public class BaseDatabase implements Serializable {

  private final static long serialVersionUID = -6805001319856652238L;
  /**
   * The arangodb host
   */
  @JsonProperty( "host" )
  @JsonPropertyDescription( "The arangodb host" )
  private String host = "127.0.0.1";
  /**
   * The arangodb server port
   */
  @JsonProperty( "port" )
  @JsonPropertyDescription( "The arangodb server port" )
  @DecimalMin( "1" )
  @DecimalMax( "65535" )
  private BigInteger port = new BigInteger( "8529" );
  /**
   * Enable tls connection
   */
  @JsonProperty( "secure" )
  @JsonPropertyDescription( "Enable tls connection" )
  private Boolean secure = false;
  /**
   * The username to login to the arangodb server
   */
  @JsonProperty( "username" )
  @JsonPropertyDescription( "The username to login to the arangodb server" )
  private String username;
  /**
   * The password of the given username
   */
  @JsonProperty( "password" )
  @JsonPropertyDescription( "The password of the given username" )
  private String password;

  /**
   * The arangodb host
   */
  @JsonProperty( "host" )
  public Optional< String > getHost( ) {
    return Optional.ofNullable( host );
  }

  /**
   * The arangodb host
   */
  @JsonProperty( "host" )
  public void setHost( String host ) {
    this.host = host;
  }

  /**
   * The arangodb server port
   */
  @JsonProperty( "port" )
  public Optional< BigInteger > getPort( ) {
    return Optional.ofNullable( port );
  }

  /**
   * The arangodb server port
   */
  @JsonProperty( "port" )
  public void setPort( BigInteger port ) {
    this.port = port;
  }

  /**
   * Enable tls connection
   */
  @JsonProperty( "secure" )
  public Optional< Boolean > getSecure( ) {
    return Optional.ofNullable( secure );
  }

  /**
   * Enable tls connection
   */
  @JsonProperty( "secure" )
  public void setSecure( Boolean secure ) {
    this.secure = secure;
  }

  /**
   * The username to login to the arangodb server
   */
  @JsonProperty( "username" )
  public Optional< String > getUsername( ) {
    return Optional.ofNullable( username );
  }

  /**
   * The username to login to the arangodb server
   */
  @JsonProperty( "username" )
  public void setUsername( String username ) {
    this.username = username;
  }

  /**
   * The password of the given username
   */
  @JsonProperty( "password" )
  public Optional< String > getPassword( ) {
    return Optional.ofNullable( password );
  }

  /**
   * The password of the given username
   */
  @JsonProperty( "password" )
  public void setPassword( String password ) {
    this.password = password;
  }

}

File: dracpond/src/main/kotlin/cc/chordflower/dracpond/application/config/base/BaseDatabase.kt
package cc.chordflower.dracpond.application.config.base

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonPropertyDescription
import com.fasterxml.jackson.annotation.JsonPropertyOrder
import io.vavr.control.Option
import jakarta.validation.constraints.DecimalMax
import jakarta.validation.constraints.DecimalMin
import java.io.Serializable
import java.math.BigInteger

/**
 * Definitions related to the arango database
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonPropertyOrder("host", "port", "secure", "username", "password")
class BaseDatabase(
  /**
   * The arangodb host
   */
  @JsonProperty("host")
  @JsonPropertyDescription("The arangodb host")
  private var host: String,
  /**
   * The arangodb server port
   */
  @JsonProperty("port")
  @JsonPropertyDescription("The arangodb server port")
  private var port: @DecimalMin(value = "1") @DecimalMax(value = "65535") BigInteger,
  /**
   * Enable tls connection
   */
  @JsonProperty("secure")
  @JsonPropertyDescription("Enable tls connection")
  private var secure: Boolean,
  /**
   * The username to login to the arangodb server
   */
  @JsonProperty("username")
  @JsonPropertyDescription("The username to login to the arangodb server")
  private var username: String?,
  /**
   * The password of the given username
   */
  @JsonProperty("password")
  @JsonPropertyDescription("The password of the given username")
  private var password: String?
) : Serializable {

  constructor(): this("localhost", BigInteger("2345"), false, null, null)

  /**
   * The arangodb host
   */
  @JsonProperty("host")
  fun getHost(): Option<String> {
    return Option.of(host)
  }

  /**
   * The arangodb host
   */
  @JsonProperty("host")
  fun setHost(host: String) {
    this.host = host
  }

  /**
   * The arangodb server port
   */
  @JsonProperty("port")
  fun getPort(): Option<BigInteger> {
    return Option.of(port)
  }

  /**
   * The arangodb server port
   */
  @JsonProperty("port")
  fun setPort(port: BigInteger) {
    this.port = port
  }

  /**
   * Enable tls connection
   */
  @JsonProperty("secure")
  fun getSecure(): Option<Boolean> {
    return Option.of(secure)
  }

  /**
   * Enable tls connection
   */
  @JsonProperty("secure")
  fun setSecure(secure: Boolean) {
    this.secure = secure
  }

  /**
   * The username to login to the arangodb server
   */
  @JsonProperty("username")
  fun getUsername(): Option<String> {
    return Option.of(username)
  }

  /**
   * The username to login to the arangodb server
   */
  @JsonProperty("username")
  fun setUsername(username: String) {
    this.username = username
  }

  /**
   * The password of the given username
   */
  @JsonProperty("password")
  fun getPassword(): Option<String> {
    return Option.of(password)
  }

  /**
   * The password of the given username
   */
  @JsonProperty("password")
  fun setPassword(password: String) {
    this.password = password
  }

  override fun equals(other: Any?): Boolean {
    if(this === other) return true
    if(javaClass != other?.javaClass) return false

    other as BaseDatabase

    if(host != other.host) return false
    if(port != other.port) return false
    if(secure != other.secure) return false
    if(username != other.username) return false
    if(password != other.password) return false

    return true
  }

  override fun hashCode(): Int {
    var result = host.hashCode()
    result = 31 * result + port.hashCode()
    result = 31 * result + secure.hashCode()
    result = 31 * result + (username?.hashCode() ?: 0)
    result = 31 * result + (password?.hashCode() ?: 0)
    return result
  }

  override fun toString(): String {
    return """{"host"="$host", "port"=$port, "secure"=$secure, "username"="$username", "password"="$password"}"""
  }


  companion object {
    private const val serialVersionUID = -6805001319856652238L
  }
}
